In [5]:
import pandas as pd
import numpy as np

# Read the CSV file
file_path = "id_ref_year_nCitation_docType.csv"  # Replace with your actual file path
df = pd.read_csv(file_path)
#df = pd.read_csv(file_path,nrows=1000)

C:\Users\ngchn\AppData\Local\Temp\ipykernel_1580\2771660964.py:6: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path)


In [7]:
# Count occurrences of each reference
ref_counts = df['ref'].value_counts()

# Create a new column with the counts
df['n_citation_ref_count'] = df['ref'].map(ref_counts).fillna(0).astype(int)


# Find which rows have changed based on the new column
changed_ids = df[df['n_citation_ref_count'] != df['n_citation_ref']]

In [16]:
changed_ids

Unnamed: 0          id         ref    year  year_ref    doc_type  \
0                 0        1091  2005687710     0.0       0.0        none   
1                 1        1091  2018037215     0.0       0.0        none   
2                 2        1388           0     0.0       0.0        none   
3                 3        1674  1535888970  2003.0    2003.0  Conference   
4                 4        1674  1992876689     0.0       0.0        none   
...             ...         ...         ...     ...       ...         ...   
8715395     8715395  1991505830  2222634690     0.0       0.0        none   
8715396     8715396  1991505830  2246658011     0.0       0.0        none   
8715397     8715397  1991505830  2249925049     0.0       0.0        none   
8715398     8715398  1991506009  2135946775     0.0       0.0        none   
8715399     8715399  1991506009  2146384916     0.0       0.0        none   

         n_citation_ref  n_citation  n_citation_ref_count  
0                     3           1                     3  
1                    12           1                    12  
2                     0           1                     0  
3                    63           1                    12  
4                     5           1                     5  
...                 ...         ...                   ...  
8715395              17          61                    17  
8715396              18          61                    18  
8715397              46          61                    46  
8715398               1           1                     1  
8715399              12           1                    12  

[8706347 rows x 9 columns]

In [17]:
# Update n_citation_ref based on the condition
changed_ids.loc[(df['n_citation_ref'] == 0) & (changed_ids['n_citation_ref_count'] != 0), 'n_citation_ref'] = changed_ids['n_citation_ref_count']


In [18]:
changed_ids

Unnamed: 0          id         ref    year  year_ref    doc_type  \
0                 0        1091  2005687710     0.0       0.0        none   
1                 1        1091  2018037215     0.0       0.0        none   
2                 2        1388           0     0.0       0.0        none   
3                 3        1674  1535888970  2003.0    2003.0  Conference   
4                 4        1674  1992876689     0.0       0.0        none   
...             ...         ...         ...     ...       ...         ...   
8715395     8715395  1991505830  2222634690     0.0       0.0        none   
8715396     8715396  1991505830  2246658011     0.0       0.0        none   
8715397     8715397  1991505830  2249925049     0.0       0.0        none   
8715398     8715398  1991506009  2135946775     0.0       0.0        none   
8715399     8715399  1991506009  2146384916     0.0       0.0        none   

         n_citation_ref  n_citation  n_citation_ref_count  
0                     3           1                     3  
1                    12           1                    12  
2                     0           1                     0  
3                    63           1                    12  
4                     5           1                     5  
...                 ...         ...                   ...  
8715395              17          61                    17  
8715396              18          61                    18  
8715397              46          61                    46  
8715398               1           1                     1  
8715399              12           1                    12  

[8706347 rows x 9 columns]

In [19]:
changed_ids.loc[changed_ids['ref'] == 0, 'n_citation_ref_count'] = 0
changed_ids.loc[changed_ids['ref'] == 0, 'n_citation_ref'] = 0


In [27]:
changed_ids.drop('Unnamed: 0', axis=1)

id         ref    year  year_ref    doc_type  n_citation_ref  \
0              1091  2005687710     0.0       0.0        none               3   
1              1091  2018037215     0.0       0.0        none              12   
2              1388           0     0.0       0.0        none               0   
3              1674  1535888970  2003.0    2003.0  Conference              63   
4              1674  1992876689     0.0       0.0        none               5   
...             ...         ...     ...       ...         ...             ...   
8715395  1991505830  2222634690     0.0       0.0        none              17   
8715396  1991505830  2246658011     0.0       0.0        none              18   
8715397  1991505830  2249925049     0.0       0.0        none              46   
8715398  1991506009  2135946775     0.0       0.0        none               1   
8715399  1991506009  2146384916     0.0       0.0        none              12   

         n_citation  n_citation_ref_count   Point  
0                 1                     3    32.0  
1                 1                    12   122.0  
2                 1                     0     2.0  
3                 1                    12  2662.0  
4                 1                     5    52.0  
...             ...                   ...     ...  
8715395          61                    17   292.0  
8715396          61                    18   302.0  
8715397          61                    46   582.0  
8715398           1                     1    12.0  
8715399           1                    12   122.0  

[8706347 rows x 9 columns]

In [22]:
changed_ids.to_csv('final.csv', index=False)

In [2]:
import pandas as pd
changed_ids = pd.read_csv("final.csv")

C:\Users\ngchn\AppData\Local\Temp\ipykernel_1580\3296430367.py:2: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  changed_ids = pd.read_csv("final.csv")


In [23]:
changed_ids

Unnamed: 0          id         ref    year  year_ref    doc_type  \
0                 0        1091  2005687710     0.0       0.0        none   
1                 1        1091  2018037215     0.0       0.0        none   
2                 2        1388           0     0.0       0.0        none   
3                 3        1674  1535888970  2003.0    2003.0  Conference   
4                 4        1674  1992876689     0.0       0.0        none   
...             ...         ...         ...     ...       ...         ...   
8715395     8715395  1991505830  2222634690     0.0       0.0        none   
8715396     8715396  1991505830  2246658011     0.0       0.0        none   
8715397     8715397  1991505830  2249925049     0.0       0.0        none   
8715398     8715398  1991506009  2135946775     0.0       0.0        none   
8715399     8715399  1991506009  2146384916     0.0       0.0        none   

         n_citation_ref  n_citation  n_citation_ref_count  
0                     3           1                     3  
1                    12           1                    12  
2                     0           1                     0  
3                    63           1                    12  
4                     5           1                     5  
...                 ...         ...                   ...  
8715395              17          61                    17  
8715396              18          61                    18  
8715397              46          61                    46  
8715398               1           1                     1  
8715399              12           1                    12  

[8706347 rows x 9 columns]

In [24]:
# Function to calculate the point
def calculate_point(row):
    if row['year_ref'] != 0:
        return (row['year_ref'] - 1800) * 10 + row['n_citation_ref'] * 10 + row['n_citation'] * 2
    else:
        return row['n_citation_ref'] * 10 + row['n_citation'] * 2

# Applying the function to calculate points
changed_ids['Point'] = changed_ids.apply(calculate_point, axis=1)

C:\Users\ngchn\AppData\Local\Temp\ipykernel_1580\2534306900.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  changed_ids['Point'] = changed_ids.apply(calculate_point, axis=1)


In [29]:
changed_ids

Unnamed: 0          id         ref    year  year_ref    doc_type  \
0                 0        1091  2005687710     0.0       0.0        none   
1                 1        1091  2018037215     0.0       0.0        none   
2                 2        1388           0     0.0       0.0        none   
3                 3        1674  1535888970  2003.0    2003.0  Conference   
4                 4        1674  1992876689     0.0       0.0        none   
...             ...         ...         ...     ...       ...         ...   
8715395     8715395  1991505830  2222634690     0.0       0.0        none   
8715396     8715396  1991505830  2246658011     0.0       0.0        none   
8715397     8715397  1991505830  2249925049     0.0       0.0        none   
8715398     8715398  1991506009  2135946775     0.0       0.0        none   
8715399     8715399  1991506009  2146384916     0.0       0.0        none   

         n_citation_ref  n_citation  n_citation_ref_count   Point  
0                     3           1                     3    32.0  
1                    12           1                    12   122.0  
2                     0           1                     0     2.0  
3                    63           1                    12  2662.0  
4                     5           1                     5    52.0  
...                 ...         ...                   ...     ...  
8715395              17          61                    17   292.0  
8715396              18          61                    18   302.0  
8715397              46          61                    46   582.0  
8715398               1           1                     1    12.0  
8715399              12           1                    12   122.0  

[8706347 rows x 10 columns]

In [30]:
changed_ids.drop('Unnamed: 0', axis=1)

id         ref    year  year_ref    doc_type  n_citation_ref  \
0              1091  2005687710     0.0       0.0        none               3   
1              1091  2018037215     0.0       0.0        none              12   
2              1388           0     0.0       0.0        none               0   
3              1674  1535888970  2003.0    2003.0  Conference              63   
4              1674  1992876689     0.0       0.0        none               5   
...             ...         ...     ...       ...         ...             ...   
8715395  1991505830  2222634690     0.0       0.0        none              17   
8715396  1991505830  2246658011     0.0       0.0        none              18   
8715397  1991505830  2249925049     0.0       0.0        none              46   
8715398  1991506009  2135946775     0.0       0.0        none               1   
8715399  1991506009  2146384916     0.0       0.0        none              12   

         n_citation  n_citation_ref_count   Point  
0                 1                     3    32.0  
1                 1                    12   122.0  
2                 1                     0     2.0  
3                 1                    12  2662.0  
4                 1                     5    52.0  
...             ...                   ...     ...  
8715395          61                    17   292.0  
8715396          61                    18   302.0  
8715397          61                    46   582.0  
8715398           1                     1    12.0  
8715399           1                    12   122.0  

[8706347 rows x 9 columns]

In [26]:
changed_ids['Point'].max()

78604.0

In [32]:
changed_ids.to_csv('point.csv',index=false)

In [2]:
import pandas as pd
import numpy as np

# Read the CSV file
file_path = "point.csv,index=false"  # Replace with your actual file path
df = pd.read_csv(file_path)
#df = pd.read_csv(file_path,nrows=1000)

C:\Users\ngchn\AppData\Local\Temp\ipykernel_20864\940130276.py:6: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path)


In [3]:
df

Unnamed: 0.1  Unnamed: 0          id         ref    year  year_ref  \
0                   0           0        1091  2005687710     0.0       0.0   
1                   1           1        1091  2018037215     0.0       0.0   
2                   2           2        1388           0     0.0       0.0   
3                   3           3        1674  1535888970  2003.0    2003.0   
4                   4           4        1674  1992876689     0.0       0.0   
...               ...         ...         ...         ...     ...       ...   
8706342       8715395     8715395  1991505830  2222634690     0.0       0.0   
8706343       8715396     8715396  1991505830  2246658011     0.0       0.0   
8706344       8715397     8715397  1991505830  2249925049     0.0       0.0   
8706345       8715398     8715398  1991506009  2135946775     0.0       0.0   
8706346       8715399     8715399  1991506009  2146384916     0.0       0.0   

           doc_type  n_citation_ref  n_citation  n_citation_ref_count   Point  
0              none               3           1                     3    32.0  
1              none              12           1                    12   122.0  
2              none               0           1                     0     2.0  
3        Conference              63           1                    12  2662.0  
4              none               5           1                     5    52.0  
...             ...             ...         ...                   ...     ...  
8706342        none              17          61                    17   292.0  
8706343        none              18          61                    18   302.0  
8706344        none              46          61                    46   582.0  
8706345        none               1           1                     1    12.0  
8706346        none              12           1                    12   122.0  

[8706347 rows x 11 columns]

In [6]:
df.to_csv('point.csv',index=False)

In [1]:
import pandas as pd
import numpy as np

# Read the CSV file
file_path = "point.csv"  # Replace with your actual file path
df = pd.read_csv(file_path)

In [2]:
df

id         ref  year  year_ref    doc_type  n_citation_ref  \
0             1091  2005687710     0         0        none               3   
1             1091  2018037215     0         0        none              12   
2             1388           0     0         0        none               0   
3             1674  1535888970  2003      2003  Conference              63   
4             1674  1992876689     0         0        none               5   
...            ...         ...   ...       ...         ...             ...   
1048570  164736972  2913217271     0         0        none             197   
1048571  164738116  2102481828     0         0        none             145   
1048572  164738116  2111308925     0         0        none            1125   
1048573  164738116  2118153703     0         0        none              49   
1048574  164738116  2151103935     0         0        none            4196   

         n_citation  n_citation_ref_count  Point  
0                 1                     3     32  
1                 1                    12    122  
2                 1                     0      2  
3                 1                    12   2662  
4                 1                     5     52  
...             ...                   ...    ...  
1048570           4                   197   1978  
1048571           1                   145   1452  
1048572           1                  1125  11252  
1048573           1                    49    492  
1048574           1                  4196  41962  

[1048575 rows x 9 columns]

In [ ]:
# Xóa những dòng có 'ref' bằng 0
df_delete_ref0 = df[df['ref'] != 0]
# Lưu kết quả ra file mới
df_delete_ref0.to_csv('point_delete_ref0.csv', index=False)


In [2]:
import pandas as pd
df_delete_ref0 = pd.read_csv('point_delete_ref0.csv')

In [3]:
df_delete_ref0

id         ref  year  year_ref    doc_type  n_citation_ref  \
0            1091  2005687710     0         0        none               3   
1            1091  2018037215     0         0        none              12   
2            1674  1535888970  2003      2003  Conference              63   
3            1674  1992876689     0         0        none               5   
4            1674  1993710814     0         0        none              10   
...           ...         ...   ...       ...         ...             ...   
966410  164736972  2913217271     0         0        none             197   
966411  164738116  2102481828     0         0        none             145   
966412  164738116  2111308925     0         0        none            1125   
966413  164738116  2118153703     0         0        none              49   
966414  164738116  2151103935     0         0        none            4196   

        n_citation  n_citation_ref_count  Point  
0                1                     3     32  
1                1                    12    122  
2                1                    12   2662  
3                1                     5     52  
4                1                    10    102  
...            ...                   ...    ...  
966410           4                   197   1978  
966411           1                   145   1452  
966412           1                  1125  11252  
966413           1                    49    492  
966414           1                  4196  41962  

[966415 rows x 9 columns]

In [10]:
# Current year
current_year = 2024

# Function to calculate citations per year
def calculate_citations_per_year(row):
    if row['year_ref'] > 0:
        return row['n_citation_ref'] / (current_year - row['year_ref'])
    else:
        return 0  # Handle case for year_ref <= 0

# Add the new column to the DataFrame
df_delete_ref0['citations_per_year'] = df_delete_ref0.apply(calculate_citations_per_year, axis=1)

In [12]:
from sklearn.preprocessing import MinMaxScaler
columns_to_scale = ['year', 'year_ref', 'n_citation_ref', 'n_citation', 'citations_per_year']
scaler = MinMaxScaler()
df_delete_ref0[columns_to_scale] = scaler.fit_transform(df_delete_ref0[columns_to_scale])

In [13]:
df_delete_ref0

id         ref      year  year_ref    doc_type  n_citation_ref  \
0            1091  2005687710  0.000000  0.000000        none        0.000347   
1            1091  2018037215  0.000000  0.000000        none        0.001906   
2            1674  1535888970  0.992075  0.992075  Conference        0.010745   
3            1674  1992876689  0.000000  0.000000        none        0.000693   
4            1674  1993710814  0.000000  0.000000        none        0.001560   
...           ...         ...       ...       ...         ...             ...   
966410  164736972  2913217271  0.000000  0.000000        none        0.033969   
966411  164738116  2102481828  0.000000  0.000000        none        0.024957   
966412  164738116  2111308925  0.000000  0.000000        none        0.194801   
966413  164738116  2118153703  0.000000  0.000000        none        0.008319   
966414  164738116  2151103935  0.000000  0.000000        none        0.727036   

        n_citation  n_citation_ref_count  Point  citations_per_year  
0         0.000290                     3     32            0.000000  
1         0.000290                    12    122            0.000000  
2         0.000290                    12   2662            0.009357  
3         0.000290                     5     52            0.000000  
4         0.000290                    10    102            0.000000  
...            ...                   ...    ...                 ...  
966410    0.001159                   197   1978            0.000000  
966411    0.000290                   145   1452            0.000000  
966412    0.000290                  1125  11252            0.000000  
966413    0.000290                    49    492            0.000000  
966414    0.000290                  4196  41962            0.000000  

[966415 rows x 10 columns]

In [20]:
# Group by 'id' and prepare to split
grouped = df_delete_ref0.groupby('id')

train_data = []
test_data = []

# Iterate over each group
for name, group in grouped:
    total_refs = len(group)
    
    if total_refs == 1:
        # If there's only one ref, add it to train
        train_data.append(group)
    elif total_refs == 2:
        # If there are two refs, add one to train and one to test
        train_data.append(group.iloc[[0]])  # First ref for training
        test_data.append(group.iloc[[1]])    # Second ref for testing
    else:
        # If there are more than two refs
        train_size = total_refs // 3  # 1/3 for training
        test_size = total_refs - train_size  # 2/3 for testing

        # Assign to train and test
        train_data.append(group.iloc[:train_size])  # First 1/3 for training
        test_data.append(group.iloc[train_size:])   # Remaining 2/3 for testing

# Combine train and test data
train_df = pd.concat(train_data).reset_index(drop=True)
test_df = pd.concat(test_data).reset_index(drop=True)

In [21]:
train_df

id         ref      year  year_ref    doc_type  n_citation_ref  \
0            1091  2005687710  0.000000  0.000000        none        0.000347   
1            1674  1535888970  0.992075  0.992075  Conference        0.010745   
2            1674  1992876689  0.000000  0.000000        none        0.000693   
3            1674  1993710814  0.000000  0.000000        none        0.001560   
4            1674  2035653341  0.000000  0.000000        none        0.001906   
...           ...         ...       ...       ...         ...             ...   
303168  164734591  1996793998  0.000000  0.000000        none        0.000000   
303169  164736972   117400678  0.000000  0.000000        none        0.000000   
303170  164736972   188145618  0.994056  0.994056  Conference        0.042808   
303171  164736972  1514298587  0.000000  0.000000        none        0.000693   
303172  164738116  2102481828  0.000000  0.000000        none        0.024957   

        n_citation  n_citation_ref_count  Point  citations_per_year  
0         0.000290                     3     32            0.000000  
1         0.000290                    12   2662            0.009357  
2         0.000290                     5     52            0.000000  
3         0.000290                    10    102            0.000000  
4         0.000290                    12    122            0.000000  
...            ...                   ...    ...                 ...  
303168    0.000580                     1     14            0.000000  
303169    0.001159                     1     18            0.000000  
303170    0.001159                    54   4558            0.045501  
303171    0.001159                     5     58            0.000000  
303172    0.000290                   145   1452            0.000000  

[303173 rows x 10 columns]

In [24]:
train_df.to_csv('train_data.csv',index=False)

In [22]:
test_df

id         ref  year  year_ref doc_type  n_citation_ref  \
0            1091  2018037215   0.0       0.0     none        0.001906   
1            1674  2087778487   0.0       0.0     none        0.000000   
2            1674  2094478046   0.0       0.0     none        0.003120   
3            1674  2100468662   0.0       0.0     none        0.010225   
4            1674  2110331535   0.0       0.0     none        0.000000   
...           ...         ...   ...       ...      ...             ...   
663237  164736972  2130521512   0.0       0.0     none        0.010919   
663238  164736972  2913217271   0.0       0.0     none        0.033969   
663239  164738116  2111308925   0.0       0.0     none        0.194801   
663240  164738116  2118153703   0.0       0.0     none        0.008319   
663241  164738116  2151103935   0.0       0.0     none        0.727036   

        n_citation  n_citation_ref_count  Point  citations_per_year  
0         0.000290                    12    122                 0.0  
1         0.000290                     1     12                 0.0  
2         0.000290                    19    192                 0.0  
3         0.000290                    60    602                 0.0  
4         0.000290                     1     12                 0.0  
...            ...                   ...    ...                 ...  
663237    0.001159                    64    648                 0.0  
663238    0.001159                   197   1978                 0.0  
663239    0.000290                  1125  11252                 0.0  
663240    0.000290                    49    492                 0.0  
663241    0.000290                  4196  41962                 0.0  

[663242 rows x 10 columns]

In [25]:
test_df.to_csv('test_data.csv',index=False)

ở các dòng nếu id giống nhau nghĩa là nó cùng một bài báo, các dòng là ref tương ứng với id đó. nếu id tiếp theo giống id trước thì nó vẫn là ref của id đó. giờ hãy phân chia số lượng bài báo lại chia thành 3 phần (1 và 2). Ví dụ bài báo id 1091 có 3 bài ref vậy thì chia data train 1 bài ref của id 1091, 2 bài ref của id 1091 là data test

In [6]:
# Finding min and max year, ignoring the default '0' year for id 1388
valid_years = df_delete_ref0[df_delete_ref0['year_ref'] > 0]['year_ref']  # Filter out years that are 0

min_year = valid_years.min() if not valid_years.empty else None
max_year = valid_years.max() if not valid_years.empty else None

# Adjust the minimum year by subtracting 0 (no change)
adjusted_min_year = min_year - 0 if min_year is not None else None

# Display the results
print("Minimum Year (adjusted):", adjusted_min_year)
print("Maximum Year:", max_year)

Minimum Year (adjusted): 1800
Maximum Year: 2019


In [5]:
#group df_delete_ref0 arcording to id and count how many ref ever id have
df_delete_ref0_count = df_delete_ref0.groupby(['id']).count()
#reset index
df_delete_ref0_count = df_delete_ref0_count.reset_index()
#rename column ref to ref_count
df_delete_ref0_count = df_delete_ref0_count.rename(columns={'ref': 'ref_count'})
#sort by ref_count
df_delete_ref0_count = df_delete_ref0_count.sort_values(by=['ref_count'], ascending=False)
#save to csv
df_delete_ref0_count.to_csv('df_delete_ref0_count.csv', index=False)

id  ref_count  year  year_ref  doc_type  n_citation_ref  \
17331   28339001        277   277       277       277             277   
84867  140172734        270   270       270       270             270   
9219    14941018        242   242       242       242             242   
41171   67710644        234   234       234       234             234   
60386   99699946        209   209       209       209             209   
...          ...        ...   ...       ...       ...             ...   
47267   77844967          1     1         1         1               1   
92678  153188475          1     1         1         1               1   
17816   29092647          1     1         1         1               1   
81455  134595198          1     1         1         1               1   
72092  119049840          1     1         1         1               1   

       n_citation  n_citation_ref_count  Point  
17331         277                   277    277  
84867         270                   270    270  
9219          242                   242    242  
41171         234                   234    234  
60386         209                   209    209  
...           ...                   ...    ...  
47267           1                     1      1  
92678           1                     1      1  
17816           1                     1      1  
81455           1                     1      1  
72092           1                     1      1  

[99683 rows x 9 columns]

In [6]:
df_delete_ref0_count.to_csv('df_delete_ref0_count.csv', index=False)

In [17]:
# Group by 'ref_count' and count the number of ids for each ref_count
df_delete_ref0_count_arcording_ref_count = (
    df_delete_ref0_count.groupby('ref_count')
    .agg(count=('id', 'count'))  # Count the number of 'id's for each ref_count
    .reset_index()
)
df_delete_ref0_count_arcording_ref_count

ref_count  count
0            1   7550
1            2   7421
2            3   7444
3            4   7158
4            5   6720
..         ...    ...
122        209      1
123        234      1
124        242      1
125        270      1
126        277      1

[127 rows x 2 columns]

In [19]:
df_delete_ref0_count_arcording_ref_count.to_csv('df_delete_ref0_count_arcording_ref_count.csv', index=False)

In [1]:

import pandas as pd
df_final = pd.read_csv('point_delete_ref0.csv')

# Đưa về input EIHI

In [4]:
# Group by 'id' and aggregate the references and points
result = (
    df_final.groupby('id')
    .agg({
        'ref': lambda x: ' '.join(map(str, x)),  # Use space instead of comma
        'Point': lambda x: f"{x.sum()}:{' '.join(map(str, x))}"  # Format with space
    })
    .reset_index()
)

# Prepare the output format
output_lines = [
    f"{row['ref']}:{row['Point']}" for _, row in result.iterrows()
]

# Save to a .txt file
with open('df_final.txt', 'w') as file:
    for line in output_lines:
        file.write(line + '\n')

print("Data has been saved to 'df_final.txt'.")

Data has been saved to 'df_final.txt'.


In [13]:
# Group by 'id' and aggregate the references and points
result = (
    df.groupby('id')
    .agg({
        'ref': lambda x: ','.join(map(str, x)),
        'Point': lambda x: f"{x.sum()}:{','.join(map(str, x))}"
    })
    .reset_index()
)

# Combine the output in the desired format with ID
result['output_with_id'] = result.apply(lambda row: f"{row['id']}:{row['ref']}:{row['Point']}", axis=1)

# Save the output with ID to a .txt file
with open('output_with_id.txt', 'w') as f:
    f.write('\n'.join(result['output_with_id']))

print("Data with IDs has been saved in 'output_with_id.txt'.")

Data with IDs has been saved in 'output_with_id.txt'.


In [17]:
from sklearn.model_selection import train_test_split

# Step 1: Read data from output.txt
with open('output_citation.txt', 'r') as file:
    lines = file.readlines()

# Step 2: Split the data into training and testing sets
train_lines, test_lines = train_test_split(lines, test_size=0.2, random_state=42)

# Step 3: Save the training and testing sets to separate files
with open('train.txt', 'w') as train_file:
    train_file.writelines(train_lines)

with open('test.txt', 'w') as test_file:
    test_file.writelines(test_lines)

print("Training and testing data have been saved as 'train.txt' and 'test.txt'.")


Training and testing data have been saved as 'train.txt' and 'test.txt'.
